modules/component/src/core/cd-aware/strategy.ts

```diff
@@ -0,0 +1,180 @@
+      : Promise.resolve();
+  return resolvedPromise;
+}
+
+function getSaveDurationSelector(): () => Observable<number> {
+  return () => defer(getResolvedPromise).pipe(mapTo(1));
+}
+
+export interface StrategyFactoryConfig {
+  component: any;
+  cdRef: ChangeDetectorRef;
+}
+
+export interface CdStrategy<T> {
+  behaviour: (cfg?: any) => MonoTypeOperatorFunction<T>;
+  render: () => void;
+  name: string;
+}
+
+export const DEFAULT_STRATEGY_NAME = 'asyncLike';
+
+export interface StrategySelection<U> {
+  asyncLike: CdStrategy<U>;
+
+  [key: string]: CdStrategy<U>;
+}
+
+export function getStrategies<T>(
+  cfg: StrategyFactoryConfig
+): StrategySelection<T> {
+  return {
+    asyncLike: createIdleStrategy<T>(cfg),
+    noop: createNoopStrategy<T>(),
+    global: createGlobalStrategy<T>(cfg),
+    local: createLocalStrategy<T>(cfg),
+  };
+}
+
+/**
+ * AsyncLike Strategy
+ *
+ * This strategy is the drop-in replacement for Angular's built-in `async` pipe.
+ * This is the only strategy that **does not also work in zone-less environments**.
+ *
+ * - \>=ViewEngine
+ * |          | Render Method      | Coalescing | Coalesce Scope |
+ * | -------- | -------------------| ---------- | -------------- |
+ * | ZoneFull | cdRef.markForCheck | ❌         | None           |
+ * | ZoneLess | cdRef.markForCheck | ❌         | None           |
+ *
+ * - Ivy
+ *
+ * |          | Render Method      | Coalescing | Coalesce Scope |
+ * | -------- | -------------------| ---------- | -------------- |
+ * | ZoneFull | cdRef.markForCheck | ❌         | None           |
+ * | ZoneLess | cdRef.markForCheck | ❌         | None           |
+ */
+export function createIdleStrategy<T>(
+  cfg: Pick<StrategyFactoryConfig, 'cdRef' | 'component'>
+): CdStrategy<T> {
+  return {
+    render: (): void => {
+      isIvy() ? markDirty(cfg.component) : cfg.cdRef.markForCheck();
+    },
+    behaviour: () => o => o,
+    name: 'asyncLike',
+  };
+}
+
+/**
+ * Noop Strategy
+ *
+ * This strategy is does nothing.
+ *
+ * - \>=ViewEngine
+ * |          | Render Method      | Coalescing | Coalesce Scope |
+ * | -------- | -------------------| ---------- | -------------- |
+ * | ZoneFull | None               | ❌         | None           |
+ * | ZoneLess | None               | ❌         | None           |
+ *
+ * - Ivy
+ *
+ * |          | Render Method      | Coalescing | Coalesce Scope |
+ * | -------- | -------------------| ---------- | -------------- |
+ * | ZoneFull | None               | ❌         | None           |
+ * | ZoneLess | None               | ❌         | None           |
+ */
+export function createNoopStrategy<T>(cfg?: any): CdStrategy<T> {
+  return {
+    render: (): void => {},
+    behaviour: () => o => o,
+    name: 'noop',
+  };
+}
+
+/**
+ *
+ * Global Strategy
+ *
+ * This strategy is leveraging the bottom up rendering approach.
+ *
+ */
+export function createGlobalStrategy<T>(
+  cfg: StrategyFactoryConfig
+): CdStrategy<T> {
+  const inIvy = isIvy();
```

Comment:
Why this strategy got isIvy() saved into a variable? (the rest of them didn't)

Reply:
Right! As isIvy is evaluated once only, I move it to a constant.